pkg/tun/tun_darwin.go

```diff
@@ -0,0 +1,247 @@
+package tun
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"runtime"
+	"unsafe"
+
+	"golang.org/x/net/ipv4"
+	"golang.org/x/net/ipv6"
+	"golang.org/x/sys/unix"
+
+	"github.com/telepresenceio/telepresence/v2/pkg/tun/buffer"
+)
+
+const sysProtoControl = 2
+const uTunOptIfName = 2
+const uTunControlName = "com.apple.net.utun_control"
+
+type Device struct {
+	*os.File
+	name string
+}
+
+// OpenTun creates a new TUN device and ensures that it is up and running.
+func OpenTun() (*Device, error) {
+	fd, err := unix.Socket(unix.AF_SYSTEM, unix.SOCK_DGRAM, sysProtoControl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open DGRAM socket: %v", err)
+	}
+	unix.CloseOnExec(fd)
+	defer func() {
+		if err != nil {
+			_ = unix.Close(fd)
+		}
+	}()
+
+	info := &unix.CtlInfo{}
+	copy(info.Name[:], uTunControlName)
+	if err = unix.IoctlCtlInfo(fd, info); err != nil {
+		return nil, fmt.Errorf("failed to getBuffer IOCTL info for %s: %w", uTunControlName, err)
+	}
+
+	if err = unix.Connect(fd, &unix.SockaddrCtl{ID: info.Id, Unit: 0}); err != nil {
+		return nil, err
+	}
+
+	if err = unix.SetNonblock(fd, true); err != nil {
+		return nil, err
+	}
+
+	name, err := unix.GetsockoptString(fd, sysProtoControl, uTunOptIfName)
+	if err != nil {
+		return nil, err
+	}
+	return &Device{File: os.NewFile(uintptr(fd), ""), name: name}, nil
+}
+
+// AddSubnet adds a subnet to this TUN device and creates a route for that subnet which
+// is associated with the device (removing the device will automatically remove the route).
+func (t *Device) AddSubnet(_ context.Context, subnet *net.IPNet) error {
+	to := make(net.IP, len(subnet.IP))
+	copy(to, subnet.IP)
+	to[len(to)-1] = 1
+	if err := t.setAddr(subnet, to); err != nil {
+		return err
+	}
+	return withRouteSocket(func(s int) error {
+		return t.routeAdd(s, 1, subnet, to)
+	})
+}
+
+// RemoveSubnet removes a subnet from this TUN device and also removes the route for that subnet which
+// is associated with the device.
+func (t *Device) RemoveSubnet(_ context.Context, subnet *net.IPNet) error {
+	to := make(net.IP, len(subnet.IP))
+	copy(to, subnet.IP)
+	to[len(to)-1] = 1
+	if err := t.removeAddr(subnet, to); err != nil {
+		return err
+	}
+	return withRouteSocket(func(s int) error {
+		return t.routeClear(s, 1, subnet, to)
+	})
+}
+
+func (t *Device) SetMTU(mtu int) error {
+	return withSocket(unix.AF_INET, func(fd int) error {
+		var ifr unix.IfreqMTU
+		copy(ifr.Name[:], t.name)
+		ifr.MTU = int32(mtu)
+		err := unix.IoctlSetIfreqMTU(fd, &ifr)
+		if err != nil {
+			err = fmt.Errorf("set MTU on %s failed: %w", t.name, err)
+		}
+		return err
+	})
+}
+
+// Read reads as many bytes as possible into the given buffer.Data and returns the
+// number of bytes actually read
+func (t *Device) Read(into *buffer.Data) (int, error) {
```

Comment:
Having `Read` and `Write` methods that are different than `io.Reader` and `io.Writer` is confusing.  Perhaps name them `ReadPacket` and `WritePacket`?